### ISSUE TITLE
[DISCUSSION] Adopting JSON-like format as next-generation model format

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#3916]
As discussed in #3878 and #3886 , we might want a more extendable format for saving XGBoost model.

For now my plan is utilizing the `JSONReader` and `JSONWriter` implemented in `dmlc-core` to add **experimental** support for saving/loading model into Json file.  Due to the fact that related area of code is quite messy and is dangerous to change, I want to share my plan and possibly an early PR as soon as possible so that someone could point out my mistakes earlier(there will be mistakes), and we don't make duplicated work. :)

@hcho3

**Comment by hcho3** [comment]
@trivialfis Are we thinking of something like `dump_model(dump_format='json')`, except that we'd be able to both read and write the new model format? (JSON dump is currently write-only)

**Comment by trivialfis** [comment]
@hcho3 `Save()` and `Load()`, with `dmlc::JSONwriter *` instead of `dmlc::stream*` as parameter.

**Comment by hcho3** [comment]
@trivialfis Awesome! We can add it as a plug-in.

**Comment by trivialfis** [comment]
@hcho3 My plan is we replace the old binary `Save` and `Load` in a few releases, so currently I'm trying to patch to code directly.  When it's done, XGBoost will save to json by default, as long as we keep the old binary `Load` function in the code, new releases of XGBoost can still read old binary format.  WDYT?

**Comment by trivialfis** [comment]
@hcho3 Good idea.  This way we don't even need to change the interface.

**Comment by hcho3** [comment]
@trivialfis Cool. I'll look forward to your pull request.

**Comment by thvasilo** [comment]
I'd argue for text models, because binary models introduce long-term compatibility and maintenance burden, [case in point](https://stackoverflow.com/questions/44582581/xgboost-binary-format-spec), and recently there have been requests for adding additional parameters to the model that cannot be satisfied easily, I think we might be breaking compatibility. [[1]](https://discuss.xgboost.ai/t/treeparam-is-it-safe-to-add-new-member-variables/368) and I remember users asking for the number of samples in each leaf.

A JSON model would make it much easier to extend, having a binary specification with explicit versions would also help greatly, see e.g. [DataSketches](https://github.com/DataSketches/sketches-core-cpp/blob/master/kll/include/kll_sketch.hpp#L490). Other options like exporting/importing to JPMML or ONNX (not sure if they support decision trees) could be left for other repositories.

**Comment by KOLANICH** [comment]
>because binary models introduce long-term compatibility and maintenance burden

Depends on design. Usually developers make the format extendable enough using tag-size-contents pattern or just checking version.

>cannot be satisfied easily because of backward compatibility,

Then I guess we should design a new binary format with extendability in mind and deprecate the old one. We will have to have some schema anyway, so no need to use generic-purpose formats like BSON/msgpack.

>Other options like exporting/importing to JPMML or ONNX (not sure if they support decision trees) could be left for other repositories.

onnx supports decision trees, but it is a generic-purpose format optimized for arbitrary computation graphs

And I have some troubles with building it.

**Comment by hcho3** [comment]
I had someone tell me that XGBoost model has endian compatibility issue. Models trained with little-endian machine cannot be loaded onto big-endian machine. Endian compatibility is one reason to favor text based format (as long as we stick to UTF-8)

**Comment by hcho3** [comment]
Also note that LightGBM uses INI-like format for model serialization.

**Comment by KOLANICH** [comment]
>Endian compatibility is one reason to favor text based format (as long as we stick to UTF-8)

Endian-compatibility is not a big problem - we can just say that little endian should be used and transform endian on serialization. Or we can just save the endian in format and transform if it doesn't match machine endianness.

**Comment by hcho3** [comment]
@KOLANICH That's a fair point. I think a good way to proceed is to have multiple parser plugins implemented and then hold a beauty content to choose the best alternative for serialization format. The criteria are
1. It should be performant.
2. It should not introduce dependencies that makes compilation too difficult.
3. It should be easy to extend without breaking backward compatibility (models given by old XGBoost should be able to load onto new XGBoost).
4. It should be independent of endianness and architecture of machines.
5. It should be compact (as in byte size of the serialized model).

**Comment by hcho3** [comment]
Personally, I'd eliminate Protocol Buffers from consideration. I've tried it in another project ([dmlc/treelite](https://github.com/dmlc/treelite)) and it takes a long time to compile from source.

**Comment by trivialfis** [comment]
I ended up re-implementing a Json library due to it's too hard to get `dmlc::JSONReader` working (`Writer` is fine).

As for other formats, the simple facts are:

1. We have different models and I want to remain the possibility of adding new ones.
2. Only the model itself knows how to save/load its internal data.
3. There are unlimited number of formats.

So a Json representation can be both a middle and an end format.  One can transform its memory representation however they like.  This can also be viewed as a reason for re-implementing a Json library, since the one in dmlc doesn't has a concrete data structure(it's a reader and loader with a stream, not exactly a data structure).

**Comment by trivialfis** [comment]
@hcho3 I kept the binary model and it's still default.  I thought about the performance issue, and it turns out there are 3-party specifications for binary form of JSON.  Just search "binary Json" should return a lot.

**Comment by hcho3** [comment]
@trivialfis Thanks for the pointer. [BSON](https://github.com/mongodb/mongo-c-driver/tree/master/src/libbson) looks like a good choice for binary encoding of JSON, given its permissive license (Apache 2.0) and support from MongoDB.
 
Also, Microsoft/LightGBM#1083 uses multi-threading to speed up serialization of large models. I'll leave more details comments in #3930 about potential use of OpenMP for serializing trees in parallel.

**Comment by hcho3** [comment]
@trivialfis [MessagePack](https://github.com/msgpack/msgpack-c) also looks like a good one as well.

**Comment by trivialfis** [comment]
Okay, I will focus the discussion in this issue rather than #3930 .  My proposal is 
1. Implement a format independent data structure storing our models in memory during serialization as the middle form.  Copied from #3930 : 

> During serialization, we will have a middle form of our models. In this PR, it's the Json class, which can be easily translated into our actual models with all parameters and model specific data preserved. Adding new parameters/models/data won't cause trouble since the class that make the changes will know how to deal with the changes by itself. While the old binary dump doesn't have such representation, to add new field, one have to hack the learner model parameter in a very careful way, and there's no way to propagate loaded data except for string parameters to other classes.
> 
> As for how the Json class is stored in disk doesn't effect Json class itself, you can traversal Json's nodes tree and obtain a complete different output format, say BSON. What I meant, specifically, is this Json class.

#3930 is already a working example (albeit Linux only currently with MSVC not building), you can use the patch from it to generate a json file to see how's it look like and room for improvement.

2. Let's first introduce plain utf-8 text Json as the end format, then consider other formats later.  From this, we can see how's it working, if it's worth the effort, possibility of new troubles or whether the extendibility is indeed improved.  Since once we have a middle form representation, implementing other formats are just items on the check list.  Anyone who feel the need can implement it.

**Comment by hcho3** [comment]
> Implement a format independent data structure storing our models in memory during serialization as the middle form

In other words, we want nested key-value stores as intermediate representation. We don't need full JSON for this purpose. Let me try to come up with a simpler alternative by tomorrow.

The idea is to 1) have all components generate nested key-value stores and 2) have the serializer convert it into the end format, such as JSON. This design ensures that we can easily switch the end format later (JSON -> MSGPACK). For 2), we can use a 3rd-party library.

**Comment by trivialfis** [comment]
@hcho3 Nice.

**Comment by hcho3** [comment]
@trivialfis BTW, good job with RTTI. I find it educational and edifying. This will come in handy for the nested key-value store I'm designing.

**Comment by hcho3** [comment]
> additional burden of type checking and error handling

@KOLANICH To be this is a feature, not a bug. Right now, the parser will happily read whatever the binary stream will provide. We do have some sanity checks (num_node should not be negative etc) but there are many holes. With the new proposal  

> generic-purpose schemaless serialization, but our data is schemeful, and it depends on native compiled code

But we want to be able to add new fields in the future without breaking backward compatibility. I am choosing JSON (or its binary encoding), because I don't want to deal with the hassle of custom-designing binary format.

**Comment by trivialfis** [comment]
@hcho3 Nice and thanks, if you were to carry out the implementation, I will try to see if it's possible to merge some parts of #3930 into the whole restructuring, it's already a working parser/dumper with a full blown key-value structure.  Please keep me posted. :)

@KOLANICH It all comes down to less fuss in the future.  Personally I want to spend more time on algorithms instead of these structuring bug fixing backward compatibility drama.  So I will prefer any idea that brings less headache for future.

**Comment by KOLANICH** [comment]
>With the new proposal, we're forced to handle errors.

But the set of errors would be much wider. Now the errors you have to process are range checking, but with a generic-purpose format you will have to check for types, names (assumming that the format is designed the way to put burden of range checking on serialization lib), that types match names and their order (for a binary format this comes for free, you just parse what you expect).

For me range-checking looks like an easier task than that.

**Comment by trivialfis** [comment]
> ut with a generic-purpose format you will have to check for types, names (assumming that the format is designed the way to put burden of range checking on serialization lib), that types match names and their order (for a binary format this comes for free, you just parse what you expect).

That's the good thing of having a algorithmic structure, please see #3930 , all the checks are seamlessly built into the structure.  Anything wrong will be caught.  For example, if you have the wrong type, you won't be able to convert it into actual value since the type information is encoded into the structure by RTTI implementation.

**Comment by hcho3** [comment]
> But the set of errors you will have to check for would be much wider.

Not really. Each class will look for a list of names (raise error if a name is not found), and then check if the found items are right types. And the order of fields doesn't matter, as JSON is un-ordered.

**Comment by KOLANICH** [comment]
> And the order of fields doesn't matter, as JSON is un-ordered.

It's true, if we parse the format entirely first, and then use it as a key-value storage, but it has some overhead related to search. There is another approach, when parsing is done on the fly and our function is called on each record, then the arguments would likely be processed in the order they are enumerated in the file. The argument about ordering has been accepted.

**Comment by trivialfis** [comment]
@KOLANICH 
> but it has some overhead related to search
Most implementation of `std::map` is red black tree, searching is performed in `O(logn)`.  In XGBoost each nested level should contain less than 10 entries, with this known bounded `n <= 10`, it's actually `O(1)`.  The overhead is a non-issue.

> when parsing is done on the fly and our function is called on each record

That will be a whole rewrite of XGBoost.  Dealing with large data blob, the better approach is to utilize `B-Tree` like data format, for example `hdf5`.  We don't need to concern about it ourselves.

**Comment by hcho3** [comment]
@trivialfis And we can use `std::unordered_map` for greater efficiency: http://supercomputingblog.com/windows/ordered-map-vs-unordered-map-a-performance-study/

**Comment by trivialfis** [comment]
@hcho3 Great.  I can try to drop the `Json::Dump`, `Json::Load` and `Json::Null` tomorrow.  See if it fits.

**Comment by tqchen** [comment]
Given that we already have a binary format. I think it is fine to just go ahead and support a json version. The two version are likely going to be sufficient

**Comment by tqchen** [comment]
For storing trees, my recommendation would be not storing them as nested data structure, instead store them as adjacency list just like the nnvm graph format. It should not hard to convert back and forth

**Comment by hcho3** [comment]
@tqchen A prototype is available at #3930.

> store them as adjacency list

I'd like to hear your reasoning about using adjacency list instead of nested dictionaries.

**Comment by trivialfis** [comment]
@tqchen Can you elaborate a little bit on using graph instead of tree?

**Comment by tqchen** [comment]
The main reason for using adj-list instead of nesting tree is the ease for parsing and edit, because as the tree went deeper(say 20 depth), it is really visually see through the tree and do editing on them. Many of the visualization libraries also depend on the adj-list structure.

The general idea of adj list is that we mark nodes as ids, and record child id of each node. So we will have things like follows(index is not really necessary):

```
[
  {"index": 0, "children" : [1, 2]},
  {"index": 1, "children" : [4, 5]},
]
```
This kind of structure is easy to edit, and there is no problem of nested parsing which can create trouble in deep nests.

**Comment by tqchen** [comment]
As you may have noticed the current in-memory data structure itself is already adj-list structure(with its cleft, cright field as integer), so it is actually easier to adopt this structure in json as well

**Comment by tqchen** [comment]
since the binary format is already compact, json format is mainly aimed at readability, so maybe we do not have to this specialization. Notably, having a clear field allows parsing from another language easy and readable(e.g. js). If we want to have a compact blob with json, just zip it or use binary format.

**Comment by trivialfis** [comment]
@tqchen In current prototype, the nodes from decision trees are stored in array, which means it's flat.

The main reason for introducing JSON format is make saving parameters easier.  The current binary form disallow changes in storing parameters.  And accroding to @hcho3 , endien is also an issue in binary format.

**Comment by tqchen** [comment]
The endian issue should be solved already if we store vector of floats, see https://github.com/dmlc/dmlc-core/pull/406

I totally support using json format, though I guess we might not need third party json libaries, dmlc/json.h might work just fine. See https://github.com/dmlc/tvm/blob/master/src/lang/reflection.cc#L398

**Comment by KOLANICH** [comment]
>The main reason for introducing JSON format

IMHO is compatibility: JSON libraries are present for almost every lang, so we get an almost free (in terms of programmer's effort) parser and serializer, but Kaitai is currently limited to some of the most popular languages and doesn't support serialization currently at all.

**Comment by hcho3** [comment]
@KOLANICH The named keys are essential to allow for future addition of parameters. For me, this is the by and large the most important motivation for using JSON for saving models. Compatness can be achieved by using binary encoding of JSON (to be added later).

@tqchen I'll look into using DMLC JSON parser.

**Comment by tqchen** [comment]
Minimum dependency is indeed important and I want to emphasize it. Because nothing is free and we need to make sure the project can be easily ported into various platforms. Most of json libraries assume schema-less model, which adds overhead. In our case, we do have a schema in the model, in which case dmlc's json parser is sufficient to do the job

**Comment by tqchen** [comment]
Given that format is something that will affect the usage quite a lot, maybe let us open an RFC on the json format proposal? We need to properly document it anyway @trivialfis @hcho3 thanks for taking lead in this

**Comment by hcho3** [comment]
Sure, I’ll draft a RFC in next few days. I think I have a good idea about how to achieve the desired goals without introducing 3rd party dependency.

I will close this issue once the RFC is up.

**Comment by hcho3** [comment]
Closing this now. The RFC document is available at #3980.
